File: src/lib/db/queries.ts
import { asc, desc, eq, sql } from "drizzle-orm";

import { getDb } from "@/lib/db/client";
import { insertLogSchema, insertSessionSchema } from "@/lib/db/validation";
import { getAccessCode } from "@/lib/ai/sentinel";
import {
  aiStocks,
  aiTrends,
  logs,
  sessions,
  type AiStock,
  type AiTrend,
  type NewAiStock,
  type NewAiTrend,
} from "@/lib/db/schema";

type PersistSentinelTurnInput = {
  sessionId: string;
  locale: "en" | "fi";
  userInput: string;
  assistantOutput: string;
  levelReached: number;
  success: boolean;
};

export async function persistSentinelTurn(input: PersistSentinelTurnInput) {
  const db = getDb();

  const sessionPayload = insertSessionSchema.parse({
    sessionId: input.sessionId,
    locale: input.locale,
    lastLevel: input.levelReached,
    unlocked: input.success,
    updatedAt: new Date(),
  });

  await db
    .insert(sessions)
    .values(sessionPayload)
    .onConflictDoUpdate({
      target: sessions.sessionId,
      set: {
        locale: input.locale,
        lastLevel: input.levelReached,
        unlocked: input.success,
        updatedAt: new Date(),
      },
    });

  const logPayload = insertLogSchema.parse({
    id: crypto.randomUUID(),
    sessionId: input.sessionId,
    locale: input.locale,
    userInput: input.userInput,
    assistantOutput: input.assistantOutput,
    levelReached: input.levelReached,
    success: input.success,
  });

  await db.insert(logs).values(logPayload);
}

export async function getLatestTrends(date?: string): Promise<AiTrend[]> {
  const db = getDb();
  const targetDate = date ?? new Date().toISOString().slice(0, 10);

  const rows = await db
    .select()
    .from(aiTrends)
    .where(eq(aiTrends.date, targetDate))
    .orderBy(asc(aiTrends.createdAt));

  if (rows.length > 0) {
    return rows;
  }

  // Fallback: return the most recent day that has rows
  const [latest] = await db
    .select({ date: aiTrends.date })
    .from(aiTrends)
    .orderBy(desc(aiTrends.date))
    .limit(1);

  if (!latest) {
    return [];
  }

  return db
    .select()
    .from(aiTrends)
    .where(eq(aiTrends.date, latest.date))
    .orderBy(asc(aiTrends.createdAt));
}

export async function getStockHistory(
  ticker: string,
  days = 365,
): Promise<AiStock[]> {
  const db = getDb();

  const cutoff = new Date();
  cutoff.setDate(cutoff.getDate() - days);
  const cutoffStr = cutoff.toISOString().slice(0, 10);

  return db
    .select()
    .from(aiStocks)
    .where(
      sql`${aiStocks.ticker} = ${ticker} AND ${aiStocks.date} >= ${cutoffStr}`,
    )
    .orderBy(asc(aiStocks.date));
}

export async function getAvailableTickers(): Promise<string[]> {
  const db = getDb();

  const rows = await db
    .selectDistinct({ ticker: aiStocks.ticker })
    .from(aiStocks)
    .orderBy(asc(aiStocks.ticker));

  return rows.map((r) => r.ticker);
}

export async function upsertTrends(rows: NewAiTrend[]): Promise<void> {
  if (rows.length === 0) return;
  const db = getDb();

  await db.insert(aiTrends).values(rows).onConflictDoNothing();
}

export async function upsertStocks(rows: NewAiStock[]): Promise<void> {
  if (rows.length === 0) return;
  const db = getDb();

  await db.insert(aiStocks).values(rows).onConflictDoNothing();
}

export async function getDashboardStats() {
  const db = getDb();
  const accessCode = getAccessCode();

  const [totals] = await db
    .select({
      totalAttempts: sql<number>`count(*)`,
      unlockedCount: sql<number>`sum(case when ${logs.success} = 1 then 1 else 0 end)`,
      directUnlockCount: sql<number>`sum(case when ${logs.success} = 1 and upper(trim(${logs.userInput})) = ${accessCode} then 1 else 0 end)`,
      highestLevel: sql<number>`max(${logs.levelReached})`,
    })
    .from(logs);

  const latest = await db
    .select({
      timestamp: logs.timestamp,
      levelReached: logs.levelReached,
      success: logs.success,
    })
    .from(logs)
    .where(eq(logs.success, true))
    .orderBy(desc(logs.timestamp))
    .limit(1);

  const successfulSessionCounts = db
    .select({
      sessionId: logs.sessionId,
      messageCount: sql<number>`count(*)`.as("message_count"),
    })
    .from(logs)
    .groupBy(logs.sessionId)
    .having(sql`sum(case when ${logs.success} = 1 then 1 else 0 end) > 0`)
    .as("successful_session_counts");

  const [averages] = await db
    .select({
      avgMessagesToUnlock: sql<number>`coalesce(avg(${successfulSessionCounts.messageCount}), 0)`,
    })
    .from(successfulSessionCounts);

  return {
    totalAttempts: totals?.totalAttempts ?? 0,
    unlockedCount: totals?.unlockedCount ?? 0,
    directUnlockCount: totals?.directUnlockCount ?? 0,
    highestLevel: totals?.highestLevel ?? 0,
    avgMessagesToUnlock: Number(averages?.avgMessagesToUnlock ?? 0),
    latestUnlockAt: latest[0]?.timestamp ?? null,
  };
}

File: goth/internal/db/queries.go
package db

import (
	"crypto/rand"
	"database/sql"
	"encoding/hex"
	"strings"
	"time"
)

// Stats mirrors the Next.js getDashboardStats() result
// (src/lib/db/queries.ts) field-for-field:
//   - TotalAttempts:       count(*) over logs.
//   - UnlockedCount:       count of successful log rows (NOT distinct sessions).
//   - DirectUnlockCount:   successful rows whose trimmed, uppercased input
//     equals the access code.
//   - HighestLevel:        max(level_reached), 0 when empty.
//   - AvgMessagesToUnlock: average of per-session total message counts over
//     sessions that contain at least one successful row, 0 when none.
//   - LatestUnlock:        timestamp of the most recent successful row, nil
//     when none.
type Stats struct {
	TotalAttempts       int
	UnlockedCount       int
	DirectUnlockCount   int
	HighestLevel        int
	AvgMessagesToUnlock float64
	LatestUnlock        *time.Time
}

func newID() string {
	b := make([]byte, 16)
	rand.Read(b)
	return hex.EncodeToString(b)
}

// UpsertSession ensures a session row exists and updates its state. The
// conflict clause mirrors persistSentinelTurn's onConflictDoUpdate in
// src/lib/db/queries.ts: locale, last_level, unlocked, and updated_at are
// all overwritten with the latest turn's values (unlocked is the latest
// turn's success flag, not a sticky maximum).
func UpsertSession(conn *sql.DB, sessionID, locale string, level int, unlocked bool) error {
	now := time.Now().UnixMilli()
	_, err := conn.Exec(`
		INSERT INTO sessions (session_id, locale, created_at, updated_at, last_level, unlocked)
		VALUES (?, ?, ?, ?, ?, ?)
		ON CONFLICT(session_id) DO UPDATE SET
			locale = excluded.locale,
			updated_at = excluded.updated_at,
			last_level = excluded.last_level,
			unlocked = excluded.unlocked
	`, sessionID, locale, now, now, level, boolToInt(unlocked))
	return err
}

// PersistTurn records a single sentinel exchange.
func PersistTurn(conn *sql.DB, sessionID, locale, userInput, assistantOutput string, level int, success bool) error {
	if err := UpsertSession(conn, sessionID, locale, level, success); err != nil {
		return err
	}
	_, err := conn.Exec(`
		INSERT INTO logs (id, session_id, locale, user_input, assistant_output, level_reached, success)
		VALUES (?, ?, ?, ?, ?, ?, ?)
	`, newID(), sessionID, locale, userInput, nullStr(assistantOutput), level, boolToInt(success))
	return err
}

// GetStats returns the aggregate counters with the exact Next.js
// getDashboardStats semantics. accessCode is the sentinel access code used
// for the direct-unlock comparison (uppercased before matching).
func GetStats(conn *sql.DB, accessCode string) (Stats, error) {
	var s Stats

	var unlocked, direct, highest sql.NullInt64
	err := conn.QueryRow(`
		SELECT
			COUNT(*),
			SUM(CASE WHEN success = 1 THEN 1 ELSE 0 END),
			SUM(CASE WHEN success = 1 AND UPPER(TRIM(user_input)) = ? THEN 1 ELSE 0 END),
			MAX(level_reached)
		FROM logs
	`, strings.ToUpper(accessCode)).Scan(&s.TotalAttempts, &unlocked, &direct, &highest)
	if err != nil {
		return Stats{}, err
	}
	s.UnlockedCount = int(unlocked.Int64)
	s.DirectUnlockCount = int(direct.Int64)
	s.HighestLevel = int(highest.Int64)

	var avg sql.NullFloat64
	err = conn.QueryRow(`
		SELECT COALESCE(AVG(cnt), 0) FROM (
			SELECT COUNT(*) AS cnt FROM logs
			GROUP BY session_id
			HAVING SUM(CASE WHEN success = 1 THEN 1 ELSE 0 END) > 0
		)
	`).Scan(&avg)
	if err != nil {
		return Stats{}, err
	}
	s.AvgMessagesToUnlock = avg.Float64

	var latest sql.NullInt64
	err = conn.QueryRow(`
		SELECT timestamp FROM logs WHERE success = 1 ORDER BY timestamp DESC LIMIT 1
	`).Scan(&latest)
	if err != nil && err != sql.ErrNoRows {
		return Stats{}, err
	}
	if latest.Valid {
		t := time.UnixMilli(latest.Int64)
		s.LatestUnlock = &t
	}
	return s, nil
}

func boolToInt(b bool) int {
	if b {
		return 1
	}
	return 0
}

func nullStr(s string) any {
	if s == "" {
		return nil
	}
	return s
}
